clipping_manager/templates/clipping_manager/books.html

```diff
@@ -0,0 +1,75 @@
+                    <button class="btn btn-dark shadow-none" id="btn-gallery">
+                        <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-grid-3x3-gap" viewBox="0 0 16 16" style="vertical-align: -2px">
+                        <path d="M4 2v2H2V2h2zm1 12v-2a1 1 0 0 0-1-1H2a1 1 0 0 0-1 1v2a1 1 0 0 0 1 1h2a1 1 0 0 0 1-1zm0-5V7a1 1 0 0 0-1-1H2a1 1 0 0 0-1 1v2a1 1 0 0 0 1 1h2a1 1 0 0 0 1-1zm0-5V2a1 1 0 0 0-1-1H2a1 1 0 0 0-1 1v2a1 1 0 0 0 1 1h2a1 1 0 0 0 1-1zm5 10v-2a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1v2a1 1 0 0 0 1 1h2a1 1 0 0 0 1-1zm0-5V7a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1v2a1 1 0 0 0 1 1h2a1 1 0 0 0 1-1zm0-5V2a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1v2a1 1 0 0 0 1 1h2a1 1 0 0 0 1-1zM9 2v2H7V2h2zm5 0v2h-2V2h2zM4 7v2H2V7h2zm5 0v2H7V7h2zm5 0h-2v2h2V7zM4 12v2H2v-2h2zm5 0v2H7v-2h2zm5 0v2h-2v-2h2zM12 1a1 1 0 0 0-1 1v2a1 1 0 0 0 1 1h2a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1h-2zm-1 6a1 1 0 0 1 1-1h2a1 1 0 0 1 1 1v2a1 1 0 0 1-1 1h-2a1 1 0 0 1-1-1V7zm1 4a1 1 0 0 0-1 1v2a1 1 0 0 0 1 1h2a1 1 0 0 0 1-1v-2a1 1 0 0 0-1-1h-2z"/>
+                      </svg>
+                    </button>
+                    <button class="btn btn-light shadow-none" id="btn-list">
+                        <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-list-ul" viewBox="0 0 16 16" style="vertical-align: -2px">
+                            <path fill-rule="evenodd" d="M5 11.5a.5.5 0 0 1 .5-.5h9a.5.5 0 0 1 0 1h-9a.5.5 0 0 1-.5-.5zm0-4a.5.5 0 0 1 .5-.5h9a.5.5 0 0 1 0 1h-9a.5.5 0 0 1-.5-.5zm0-4a.5.5 0 0 1 .5-.5h9a.5.5 0 0 1 0 1h-9a.5.5 0 0 1-.5-.5zm-3 1a1 1 0 1 0 0-2 1 1 0 0 0 0 2zm0 4a1 1 0 1 0 0-2 1 1 0 0 0 0 2zm0 4a1 1 0 1 0 0-2 1 1 0 0 0 0 2z"/>
+                        </svg>
+                    </button>
+                </div>
+            </div>
+        </div>
+    </div>
+    <div id="book-view" class="gallery-view">
+        <div class="list-header">
+            <div></div>
+            <div>{% trans "Title" %}</div>
+            <div>{% trans "Author" %}</div>
+            <div>{% trans "Clippings" %}</div>
+        </div>
+        {% for book in books %}
+            <div class="book-element">
+                <a href="{% url 'clipping_manager:browse'%}{% urlparams book=book.id content='' %}">
+                    <div class="view-element">
+                        <div class="book-cover">
+                            <img src="https://i.pinimg.com/736x/8d/13/b1/8d13b15006817344608393deb8b3a9a2--old-book-cover-book-covers.jpg" alt="plain book cover">
+                        </div>
+                        <div class="book-title">
+                            {{book.title|default:"unknown"}}
+                        </div>
+                        <div class="book-author">
+                            {{book.author_name|default:"unknown"}}
+                        </div>
+                        <div class="book-clippings-count">
+                            {{book.clippings.all|length}}
```

Comment:
This will lead to n+1 queries without prefetching the clipping count (which shouldn't be too much of an issue since the number of books in the list is always "small").

But getting the `.all` queryset and then using the `length` filter could be done better by simply using `clippings.count` can't it? 

If one wants to do this "nicely"/"efficiently" one could use [annotations](https://docs.djangoproject.com/en/dev/topics/db/aggregation/#generating-aggregates-for-each-item-in-a-queryset)

Reply:
Done!
Is there a radical difference between using `annotate` with `Count()` and `books.clippings.count` inside a template? Does the former use `group by` and collect all the data at once when a request comes in, and the latter must perform `join` operation on each element to get it's clippings?

Reply:
> Does the former use group by and collect all the data at once when a request comes in, and the latter must perform join operation on each element to get it's clippings?

`book.clippings` itself will already lead to `n+1` queries (even without the count; unless prefetching it), because the queryset of `books` doesn't include the clippings.

I'm not 100% sure how `book.clippings.count` will behave if you prefetch the clippings. Might be a `COUNT` SQL statement.